lib/MonadicRewrite.thy

```diff
@@ -14,24 +15,100 @@ imports
   LemmaBucket
 begin
 
-definition
-  monadic_rewrite :: "bool \<Rightarrow> bool \<Rightarrow> ('a \<Rightarrow> bool)
-                       \<Rightarrow> ('a, 'b) nondet_monad \<Rightarrow> ('a, 'b) nondet_monad \<Rightarrow> bool"
-where
- "monadic_rewrite F E P f g \<equiv> \<forall>s. P s \<and> (F \<longrightarrow> \<not> snd (f s))
-           \<longrightarrow> (E \<longrightarrow> f s = g s)
-            \<and> (\<not> E \<longrightarrow> fst (g s) \<subseteq> fst (f s) \<and> (snd (g s) \<longrightarrow> snd (f s)))"
+definition monadic_rewrite ::
+  "bool \<Rightarrow> bool \<Rightarrow> ('s \<Rightarrow> bool) \<Rightarrow> ('s, 'a) nondet_monad \<Rightarrow> ('s, 'a) nondet_monad \<Rightarrow> bool"
+  where
+   "monadic_rewrite F E P f g \<equiv>
+      \<forall>s. P s \<and> (F \<longrightarrow> \<not> snd (f s))
+          \<longrightarrow> (E \<longrightarrow> f s = g s) \<and> (\<not> E \<longrightarrow> fst (g s) \<subseteq> fst (f s) \<and> (snd (g s) \<longrightarrow> snd (f s)))"
```

Comment:
I'm not sure that's exactly the case. Attempting a proof:
```
lemma monadic_rewrite_antisym:
  "monadic_rewrite F True P f g = (monadic_rewrite F False P f g ∧ monadic_rewrite F False P g f)"
  unfolding monadic_rewrite_def
  apply (rule iffI)
  subgoal
    apply clarsimp
    apply (cases F; clarsimp)
```
This leaves us in this state:
```
⋀s. ⟦∀s. P s ∧ ¬ snd (f s) ⟶ f s = g s; P s; ¬ snd (g s); F⟧ ⟹ fst (f s) ≤ fst (g s) ∧ ¬ snd (f s)
```
Which as far as I can see doesn't have a proof. Thoughts?

Reply:
Hm, there might be something going in the wrong direction in my thinking. Let me check in Isabelle.

Reply:
To report back: anti-symmetry doesn't hold if F is true, because of the way monadic_rewrite is defined (it does for F = false).

`monadic_rewrite` (for E=false) is not actually `no_fail P f --> refined_by P f g` where `refined_by` would be the subset part of the `monadic_rewrite` definition. Instead it makes a weaker assumption on `f`, which leads to a stronger overall statement. The weaker assumption is that `f` does not have to satisfy all of `no_fail P`; instead absence of failure in `monadic_rewrite` is only relevant for states where `g` would otherwise not refine `f`.

My gut feeling is that the slightly weaker `no_fail P f --> refined_by P f g` is what we actually want, because it has nicer properties and relates nicely to existing concepts. There are more `f` and `g` for which `no_fail P f --> refined_by P f g` holds than for `monadic_rewrite True False f g`, but I doubt there is any real difference in practice.